### DISCUSSION TITLE
DS2482 1-Wire Host Controller with Tasmota

### DISCUSSION THREAD

**Discussion by augea** [arendst/Tasmota#13932] [Ideas]
I use several DS18B20 in 1-Wire network. Current they are connected to DS2482, a 1-Wire Host Controller which is best for 1-Wire timing. This 1-Wire host controller translates the 1-Wire devices to I2C.  Does it make sense to implemet this as well in Tasmota. At the moment I use this 1-host controller with a Rasberry Pi but I want to migrate to Tasmota and ESP32. I found on the internet a project for the DS2482. "https://github.com/rickkas7/DS2482-RK". Maybe this code can be implemented in Tasmota as well. 
Does it make sense to use a DS2482 with Tasmota?

**Comment by sfromis** [comment]
Sure, a better way of controlling DS18B20 could indeed make sense with Tasmota. Anyone can create a driver for DS2482 and submit a PR; this is how most supported sensors have been added. With I2C and ESP32, there's also the the Berry language, which works quite well for creating drivers for I2C devices, in an easier way than creating a driver in C++.

**Reply by augea** [reply]
I will check the Berry documentation. Maybe I'm able to wirte a script for an I2C driver. Never done this before.

**Comment by plewka** [comment]
Sounds VERY interesting for me, too. 
You probably have got the -100 in mind. I've got hardware which uses the -800 device having eight 1wire buses. With plain "ArduinoESP32" was able to tweak the code to read the sensors (don't have other devices on bus) stable over weeks. 
Latest I recognized even the latest code I found (with patches) has had issues when switching and was never tested with the -800. A few months ago I got around the problem with a delay here and there but now something changed in the base and I needed to do a change after the selection of another bus. 
I'd like to contribute code to Tasmota, since the API for the Bit-Banging implementation looks quite compatible (I was able to replace the original library in Arduino with minor changes in the constructor and for switching because of this).

My problem is: How to add it to Tasmota tin he "architectural" right way? Maybe it's Berry (which I didn't look at now), but it offers a bus (1wire) which is used by the DS18B20 sensors. Can this be done with berry in a smooth way?

I'll have to support a newer OneWire-I2C device next, too, since it is on a newer hardware. Would be nice to do it right ;-)

**Reply by sfromis** [reply]
What was suggested was a device connecting to the ESP chip via I2C, *instead* of the sometimes difficult to handle onewire bus. Tasmota has lots of devices connecting via I2C, where the driver will create JSON and also segments for the front page, both possibly including multiple devices, like if the same type of sensor is repeated. Using Berry is another avenue, which may be "more accessible" than the more complex setup needed for creating a driver in C++, but of course Berry is only available with ESP32, and most Tasmota usage does happen on ESP8266 devices.

The one-wire bus access to DS18B20 has been a source of much trouble in Tasmota, as those sensors (including clones of "varying quality") can be finicky about timing, with the same driver having had trouble getting to handle all sensors. I2C connection tends to work much better than onewire, and if there is a controller chip better at dealing with onewire sensors, great!

**Reply by plewka** [reply]
For me the current one-wire support creates limitations because of the number of sensors. That's why I wasn't too interested in past and before ESP32.

Before Corona-shortage the sensors were very cheap. I paid ~1€ for one that time. No problem to widely spread them around. I did an 8 channel heating control a few years ago which uses a multiplexer channel for every room, because I didn't want to loose all temperatures if there is a damage in one room. This is where the -800 came in. 
I've got 8 sensors in the rooms and 8 on their drainback-pipes plus two on the main flow lines. If there is no supply on one wire there are issues - but one must say it's a local bus from design to be fair. With power I didn't see any issues of the sensors.

My number is too much for the sensors string of Tasmota as far as I understand...especially if there are additional BME280 or even 680. Finally a limitation of the mqtt buffer if I understand correctly. I noticed that, too and went to secondary topics. For direct human view you don't need all these sensors in one topic. This can be solved somehow for Tasmota I'm sure.

I'll put the code which worked for me in plain Arduino to Gitlab. Maybe we find someone who is able to tell what's the right way to add that support :-)

**Comment by plewka** [comment]
If you have a look at the dallas-sensors library and how to detect original sensors from clones and a possible implementation of a check inside the code or a separate library there might be the answer why there are well and badly behaving devices ;-)

**Comment by lalo-uy** [comment]
Is not that easy to tell clones apart.

Look here:   https://github.com/cpetrich/counterfeit_DS18B20

**Comment by plewka** [comment]
I would be very thankful to hear an advice/discussion on how to optionally "replace" the bitbanging OneWire
to the I2C way. I tried to ask, but it went a helpful, but other way. Can this be done with an I2C driver
or is there an architectural conflict? 
I would like to target the DS2482-800 and the DS2485...

**Reply by barbudor** [reply]
There should be no reasons why you couldn't write an I2C driver for DS2482-100 or -800
I don't think there is an interrest for the DS2485, unless you can make a single driver for all 3 you can probbaly forget this one as the DS2482-100 seems more versatile as it supports parasitic mode.
Is there really a need for the -800 ?
As written above at some point there is a limit in the number of sensors you can report in a MQTT SENSOR message

**Comment by plewka** [comment]
I've got the DS2485 on hardware so I have got some "limited" interest ;)
Same for DS2482-800. The -100 works transparent - simply ignores the bus switching.
Regarding the limitation of the sensor string I've got the idea of splitting it and use
sensors2... Maybe as post processing before sending maybe as a second structure
for these sensors only.
Basically I'm very happy with 8, max. 10 primary sensors, the others could come on 
request in a different way. I'll have to think about that...

Since MQTT is not limited itself one could even go for a bigger buffer on lower level. 
With bigger embedded systems we're transmitting Multi-Megabyte datasets with some 
throughput via MQTT in big buffers. As long as the brooker is happy with it and small 
footprint clients don't need to subscribe it...

**Reply by plewka** [reply]
Bigger packets behave bad on weak connections I know...it's for cabled ethernet.

**Reply by sfromis** [reply]
The issue with bigger packets is not so much about connection quality, but about limited RAM leading to limited buffer sizes.

Assuming that the host controller with a new Tasmota driver gathers data from a lot of DS18B20, it would be quite feasible to have those values transmitted in a **much** shorter payload than used with the standard sensor support in Tasmota. The default format is approx 55 characters per DS18x20 sensor, if instead having an array with one element per sensor, that could be 5-7 characters per sensor (assuming `TempRes 2`), plus the sensor wrapping one time for the whole array. For clarity, it would make sense to do a one-time payload with the hex ID of which sensor is at what position in the array, and of course keep that position static.

**Comment by barbudor** [comment]
And SENSOR2 is definitively to avoid

**Reply by sfromis** [reply]
In case of splitting into more than one payload, you should have the sensor names separate (like when Tasmota makes a different name for each instance of the same type), and have them in a new payload on the same MQTT topic.

**Reply by barbudor** [reply]
I don't like having different content on the same topic, like 2 SENSOR messages, one with DS 1 to 10 and another one with DS 11 to 20
An alternative is an option to have each sensor on a separated topic like `tele/%topic%/DS18X20/<id>` like on the BLE where the MAC is used as `<id>`
This is specifically more interresting if using SensorRetain
But that means the driver as to take care of the publishing, not using the standard JSON concatenation.

**Reply by sfromis** [reply]
Those complications are why I suggested publishing sensor values in an array, when there are many sensors connected to the same controller.

**Reply by barbudor** [reply]
With DS18x20 you will still have the same problem that sensors may not come up at the same place in the array across reboots. So a dictionnay keyed with the ID is still needed.

**Reply by sfromis** [reply]
Right, as I already suggested, the dictionary could be be separate, keeping the same array index for dictionary and sensor readings.

**Comment by plewka** [comment]
A dictionary is what I thought about, and how I did it in my old design. Since the sensors are used for fixed tasks/locations and it is not helpful if there is a shift even if a new sensor comes in (f.e. as a replacement). I would say I really need a mapping. I agree "index plus temp-value" is already shorter than the full MAC-ID. 

So I won't go for an other topic, I understand. Maybe it's a solution to limit the number of 1w-sensors shown via
SENSORS and go for a second interface which uses request+answer for more sensors. 
I like to look at them from time to time, but there won't be close coupling to other Tasmota clients...only a database
log. I'll probably be _very_ happy with 3*8 absolute maximum, however.

**Comment by plewka** [comment]
How critical is a change to the json structure...for example using an array "temperatures" and save the repeated
string? Even with the IDs - maybe shortened - it would be more compact if this doesn't hurt too much:
**FROM:**
{"Time":"2021-01-02T09:09:44","DS18B20-1":{"Id":"00000566CC39","Temperature":-10.1},"DS18B20-2":{"Id":"0000059352D4","Temperature":-10.2},"DS18B20-3":{"Id":"000005937C90","Temperature":-10.3},"DS18B20-4":{"Id":"0000059352D4","Temperature":-10.4},"DS18B20-5":{"Id":"000005937C90","Temperature":-10.5},"DS18B20-6:{"Id":"0000059352D4","Temperature":-10.6},"DS18B20-7":{"Id":"000005937C90","Temperature":-10.7},"TempUnit":"C"}
**TO:**
{"Time":"2021-01-02T09:09:44","DS18B20":{"temperature":[-10.1,-10.2,-10.3,-10.4,-10.5,-10.6,-10.7]},"TempUnit":"C"}
**OR:**
{"Time":"2021-01-02T09:09:44","DS18B20":{"temperature":[-10.1,-10.2,-10.3,-10.4,-10.5,-10.6,-10.7] "id":["66CC33","66CC34","66CC35","66CC36","66CC37","66CC38","66CC39"]},"TempUnit":"C"}

**Reply by ascillato** [reply]
> How critical is a change to the json structure

Well, it is a big breaking change for a lot of users and also for several Tasmota related tools that are already using this way. We try to avoid breaking changes as much as we can. Besides the actual JSON structure was introduced a very long time ago so that change is not recommended. Sorry.

**Reply by plewka** [reply]
Ok, I understand and agree for stable APIs. How about an alternative "compressed" mode?
Would this be a legal possibility if done right?
 
Upps, when I see this, the BME680s come to my mind...didn't care for them until now. 
I'll will fail much earlier with them. Actual hardware uses 10 of them plus few DS18B20.

**Reply by barbudor** [reply]
> {"Time":"2021-01-02T09:09:44","DS18B20":{"temperature":[-10.1,-10.2,-10.3,-10.4,-10.5,-10.6,-10.7]},"TempUnit":"C"} 
Bad because you loose the order in case of reboot or when changing the number of sensors.
You must keep a fixed way of identifiying each sensor separately

> {"Time":"2021-01-02T09:09:44","DS18B20":{"temperature":[-10.1,-10.2,-10.3,-10.4,-10.5,-10.6,-10.7] "id":["66CC33","66CC34","66CC35","66CC36","66CC37","66CC38","66CC39"]},"TempUnit":"C"}
Not very good because it `doesn't leverage the easyness of dictionnary search usable in multiple languages

My preference would be:
```
{"Time":"2021-01-02T09:09:44","DS18B20":{"00000566CC39":-10.1,"0000059352D4":-10.2,"000005937C90":-10.3, "0000059352D4":-10.4,"000005937C90":-10.5,"0000059352D4":-10.6,"000005937C90":-10.7},"TempUnit":"C"}
```
This is perfectly iterable and searchable.
If you want to reduce the size of the ID, may be looking for some kind of hash function but we need to be sure it will remains unique

**Reply by sfromis** [reply]
Changing JSON structure would **only** make sense within the realm of a new driver with special needs, not as a general purpose change. While the structure with arrays would be more complex, and not really make sense for your short payloads in the examples, it would also be possible to have the dictionary in a separate payload. The complexity needed for this would probably not make sense unless having a large number of same-type sensors.

What I'd probably do instead of all this would be extra ESP8266/ESP32 boards, instead of a ton of wiring and complexity. The benefits of cramming everything onto one MCU are not large.

**Reply by plewka** [reply]
You are basically right. It even prevents electrical ground loops, noise and capacitance on the bus and so on. 
Even the ESPs are made for a more granulated concept...
I didn't go for Tasmota for this projects in past, but the benefits of Tasmota with the ESP32, are powerful and great, 
even if I have to change a few things. If I can do it in a way useful for others, I'll take that way...I know it's not the generic
living room use for sure :-)

I can't use WLAN because of radiated noise. Any further ESP requires an additional PoE-Ethernet connection.
For one of my two use cases the CPU only has to read the sensors from time to time (very slow process)  and has to switch 8 valves on and off...no own processing (I2C clock set to minimum). The update of the sensors would be allowed to be very slow, because the medium flow is slow. The process waits for the valves to react. When in use, its controlled via MQTT only. Controlled from an Embedded Linux host, visualized via Grafana.
The 10 BMEs work with some kind of topic compression - when kicking out Due point - I just counted bytes :-)
It is indeed possible to go for a second node, but no more.

**Comment by plewka** [comment]
@barbudor: This indeed looks quite attractive to me, too. Even when walking through the loop at creation, it is easier.
But isn't it a problem to fully lose "temperature" in the topic? Ok, that DS is temperature sensor and the unit is attached...
For the BMEs that's not a direct solution - would start with these. Hmm.

**Reply by barbudor** [reply]
Correct
Loosing the indication of Temperature is not a good idea

Whatever, as Adrian pointed, we should avoid to break existing
And with the current format, it is still possible to manage ~20 DS18x20 on a single MQTT message
More than that would be above what Tasmota is designed for

**Comment by plewka** [comment]
I've got an other idea...while I can probably live fine with hat number of possible DS18x20 - I understood the number to be much smaller - I'm only in trouble with the BMEs. I changed the driver to support 16 of them via an 8ch mux and used busid+nodeid instead of I2C address as -postfix (f.e. BME680-60...bus6, idx0).

Now the idea: Since the sensors give their ID each time the sensor data is send, I could go
and interleave the sensors currently send, or move a window. If you don't attach too many of them, everything is normal.
Even for 1wire it should be a possible as a "short time failure" to have one sensor missing for a while.
Would be a good idea not to crash because of that ;-)

If it's too much for one packet, a subscriber will miss a value in one packet, while it is there in the next.
If you go for more sensors, you have to care for missing sensors but this would be a good idea to
do even if you only have a hand full of sensors active, too.

What do you think?
I don't oversee what is happening to Webview f.e.... does it hurt there? 
Will pick up more of these sensors and have a practical look tomorrow.

**Reply by plewka** [reply]
@barbudor @sfromis 
When looking for the actual size limit I found a notice in xdrv_02_09_mqtt.ino which 
directed me to PubSubClient.h:
Packet size definition changed with tasmota 8.1.0.8 - much more space since then!
Enough for my 10 BMEs.
 
```
// MQTT_MAX_PACKET_SIZE : Maximum packet size. Override with setBufferSize().
#ifndef MQTT_MAX_PACKET_SIZE
//#define MQTT_MAX_PACKET_SIZE 256
#define MQTT_MAX_PACKET_SIZE 1200     // Tasmota v8.1.0.8
#endif
```

**Comment by plewka** [comment]
@augea 
Just a first feedback...
At least the code you mentioned "https://github.com/rickkas7/DS2482-RK" works directly for one channel on 
my hardware in a plain *.ino on plattform-io! 
That's much better than my experiences with other codes in past. It even claims to be done asynchronously
and doesn't wait for the sensors. This is fine! Don't know how Tasmota treats the ds18s20 with 
the bitbanging interface now (didn't have a closer look). 
RK's library is done in modern C++ for Particle.IO  - slightly ahead of my capabilities.
After I tweaked the formated prints and the log, it directly found two sensors on channel 0 of the ds2482-800 
and offers usefull values, now.

**Reply by plewka** [reply]
Select channel indeed works, too. 
Now I need to understand how to traverse through the channels...

**Reply by augea** [reply]
@plewka Thank you for yor work an testing the software.

**Comment by w2533** [comment]
@augea
Thank you for your work on a DS2482 driver. As I'm using a huge number of DS18B20 in parasitic mode, that would solve my problems.
Looking forward on the results!
Cheers Wolfgang

**Comment by augea** [comment]
For me it’s still unclear how to integrate the driver and active in the GUI /change I2C address it if necessary. I was already able to compile my own custom software but this is ath the moment beyond by knowledge. Is somebody able to finally integrate into the user_config_override or how do I merge the code with the latest release?

**Comment by plewka** [comment]
Regarding the code one has to understand everything is an object there...down to the commands on the 1W-Bus.
On first view I didn't recognize the full testing code. It does more than required. It uses an I2C Maxim to switch power to the sensors and it demonstrates the -100 an the -800. If you have a look at it, you see the sub buses are again instances of a class.
The sub-buses are stored in a list. I "fixed" the testing code to compile without particle...but no progress for Tasmota, though.

Regarding Tasmota I'm early beginner and not the right guy to explain...I can only try to do.
You take a new ID (I2C-Section) for the new driver and regarding the I2C-ID there is search-code in each driver. Best Idea is to identify the chip by registers or magic signatures if possible.
Tasmota is prepared to look at more Driver-IDs than actually in use (nothing more to do, but use the *.ino for the code). If the I2C-ID is shared with other drivers you have to add more masking in the #defines or disable the driver which conflicts at the user_config_override. 
You might even see a conflict on the webpage if a driver shows up for hardware which is not there...There is even a full documentation matrix which shows which variant (f.e. SENSORS) offers which drivers -> I2C-IDs and which has priority.
The (web)interface of the driver is offering (not only) shows up depending on the driver. I2C BME sensors - you visually define the bus interface only - appear as additional values on the main page.

At the end of the driver code there is the interfacing code to Tasmota using a Lamda interface. It dispatches the calls to the functions in the middle section of the code. The primary driver code is in the first code section, maybe supported by libraries.
Merging the code...you fork the original code and do your changes. I didn't know VisualStudioCode before, and even the GIT integration is quite useful I must say. 
From time to time you use fetch upstream in Github (Web). Usually there won't be a conflict and you end up having a fork which is ahead only, not behind. Finally you can ask for a pull to the original code - maybe it is accepted. If not, you try to enhance or follow development in parallel by fetching the changes and repair conflicts in future.

**Comment by alastaid** [comment]
Hi, resurrecting this thread as I have just replaced a Pi with an ESP32 running tasmota to control some relays.  The pi did have three DS18B20's connected on a long (30m) cable via a DS2482-100, and worked perfectly for 5 years.  Although not essential I would like to get the temperatures off the sensors via tasmota.  I am unclear from reading this thread (sorry, newbie), if there is some code I need to add in and compile my own tasmota to get this working, or has something got submitted to the main and I need to configure it, or something else?  Thanks in advance for any advice

**Reply by sfromis** [reply]
From what I can see, a Tasmota driver for the DS2482 was not created to be included in the project, but Tasmota standard builds does support 8 DS18B20 on the same bus, or up to 4 busses.

With 3 DS18B20, you may be able to get it to work connecting directly to the ESP32, as long as you have a good setup for power and pullup. Of course, direct wiring with long cables could pick up electrical noise, with the risk of affecting the MCU, depending on your environment.

**Reply by barbudor** [reply]
@alastaid 
As far as I see, the DS2482 is an I2C controller for 1-wire bus.
Quickly said, as Tasmota directly has a 1-wire driver for the DS18B20, you shouldn't need the DS2482.

Some people have been able to connect 12+ DS18B20 to a Tasmota host so I would bet this includes long lines use-cases.

**Comment by thermseekr** [comment]
Hi! Happy new year everyone!

Has this gone any further?

Thanks
Tales
